rust/src/server.rs

```diff
@@ -568,113 +608,155 @@ impl Seabird for Arc<Server> {
+                proto::ChatRequestInner::SendPrivateMessage(proto::SendPrivateMessageChatRequest {
                     user_id,
                     text: req.text,
                 }),
             )
             .await?;
 
         match resp {
-            ChatEventInner::Success(_) => Ok(Response::new(proto::PerformPrivateActionResponse {})),
+            ChatEventInner::Success(_) => Ok(Response::new(proto::SendPrivateMessageResponse {})),
             ChatEventInner::Failed(failed) => Err(Status::unknown(failed.reason)),
             _ => Err(Status::internal("unexpected chat event")),
         }
     }
 
-    async fn send_message(
+    async fn perform_action(
         &self,
-        req: Request<proto::SendMessageRequest>,
-    ) -> RpcResult<Response<proto::SendMessageResponse>> {
+        req: Request<proto::PerformActionRequest>,
+    ) -> RpcResult<Response<proto::PerformActionResponse>> {
+        let tag = extract_auth_tag(&req)?;
         let req = req.into_inner();
         let (backend_id, channel_id) = req
             .channel_id
             .parse::<FullId>()
             .map_err(|_| Status::invalid_argument("failed to parse channel_id"))?
             .into_inner();
 
+        self.broadcast(EventInner::PerformAction(proto::PerformActionEvent {
+            sender: tag.to_string(),
+            inner: Some(proto::PerformActionRequest {
+                channel_id: req.channel_id,
+                text: req.text.clone(),
+            }),
+        }));
+
         let resp = self
             .issue_request(
                 backend_id,
-                proto::ChatRequestInner::SendMessage(proto::SendMessageChatRequest {
+                proto::ChatRequestInner::PerformAction(proto::PerformActionChatRequest {
                     channel_id,
                     text: req.text,
                 }),
             )
             .await?;
 
         match resp {
-            ChatEventInner::Success(_) => Ok(Response::new(proto::SendMessageResponse {})),
+            ChatEventInner::Success(_) => Ok(Response::new(proto::PerformActionResponse {})),
             ChatEventInner::Failed(failed) => Err(Status::unknown(failed.reason)),
             _ => Err(Status::internal("unexpected chat event")),
         }
     }
 
-    async fn send_private_message(
+    async fn perform_private_action(
         &self,
-        req: Request<proto::SendPrivateMessageRequest>,
-    ) -> RpcResult<Response<proto::SendPrivateMessageResponse>> {
+        req: Request<proto::PerformPrivateActionRequest>,
+    ) -> RpcResult<Response<proto::PerformPrivateActionResponse>> {
+        let tag = extract_auth_tag(&req)?;
         let req = req.into_inner();
         let (backend_id, user_id) = req
             .user_id
             .parse::<FullId>()
             .map_err(|_| Status::invalid_argument("failed to parse user_id"))?
             .into_inner();
 
+        self.broadcast(EventInner::PerformPrivateAction(
```

Comment:
You could make these broadcasts not so repetitive with a trait and some generics, but the types would make them similarly long. It is frustrating how similar all of these methods are.

Reply:
Yeah, since we only broadcast in a few methods, I'm not quite sure it's worth it yet. Eventually, I feel like I should look at macros for this.